### PR TITLE
Improvements on Go arrays

### PR DESCRIPTION
**Pull request by h12w**
Support encoding and decoding of Go byte array in bin format (besides byte slices, byte arrays should be encoded in bin format too)

### CODE REVIEW

File: example_encodeMapStringString_test.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ func Example_encodeMapStringString() {
 		panic(err)
 	}
 
-	decodedMapValue := value.(map[interface{}]interface{})
+	decodedMapValue := value.(map[string]interface{})
```
Comment (vmihailenco) [inline example_encodeMapStringString_test.go]: We can't make such change, because it breaks existing user apps.

File: map.go [hunk 2]
```diff
@@ -68,7 +73,23 @@ func decodeMap(d *Decoder) (interface{}, error) {
...
+	for key := range m {
+		if _, isStr := key.(string); !isStr {
+			return nil, errNotStringMap
+		}
+	}
+	strMap := make(map[string]interface{})
+	for key, value := range m {
+		strMap[key.(string)] = value
+	}
+	return strMap, nil
```
Comment (vmihailenco) [inline map.go]: Please revert, because this breaks existing behaviour. We can only do this in new major version (v3).

File: typeinfo.go [hunk 3]
```diff
@@ -267,6 +267,23 @@ func encodeBytesValue(e *Encoder, v reflect.Value) error {
...
+	var bs []byte
+	if v.CanAddr() {
+		bs = v.Slice(0, v.Len()).Bytes()
+	} else {
+		bs = make([]byte, v.Len())
+		for i := range bs {
+			bs[i] = uint8(v.Index(i).Uint())
+		}
+	}
+	return e.EncodeBytes(bs)
```
Comment (vmihailenco) [inline typeinfo.go]: I see 2 possibilities:

- use reflect.Copy

```
	src := [3]byte{1, 2, 3}
	dst := make([]byte, len(src))
	n := reflect.Copy(reflect.ValueOf(dst), reflect.ValueOf(src))
	fmt.Println(n, dst)
```

I expect it to be faster, but would be nice to have a benchmark.

- don't create intermediate buffer, e.g.

```
e.EncodeBytesLen(v.Len())
for i := 0; i < v.Len(); i++ {
    e.w.WriteByte(v.Index(i).Uint())
}
```

Do you want to try it yourself? Let me know you want me to pick up this PR.

File: slice.go [hunk 4]
```diff
@@ -282,6 +285,24 @@ func (d *Decoder) skipSlice(c byte) error {
...
+	bs, err := d.getBytes(n)
+	if err != nil {
+		return err
+	}
+	for i := range bs {
+		if i >= v.Len() {
+			break
+		}
+		v.Index(i).Set(reflect.ValueOf(bs[i]))
+	}
```
Comment (vmihailenco) [inline slice.go]: Use `reflect.Copy` here?

Comment (vmihailenco) [inline slice.go]: Or alternatively:

```
for i := 0; i < n; i++ {
    c, err := d.r.ReadByte()
    if err != nil { return err }
    v.Index(i).SetUint(c)
}
```

File: slice.go [hunk 5]
```diff
@@ -282,6 +285,24 @@ func (d *Decoder) skipSlice(c byte) error {
...
+	bs, err := d.getBytes(n)
+	if err != nil {
+		return err
+	}
+	for i := range bs {
+		if i >= v.Len() {
+			break
+		}
+		v.Index(i).Set(reflect.ValueOf(bs[i]))
```
Comment (vmihailenco) [inline slice.go]: SetUint?

**Comment by h12w** [comment]
1. Incompatible changes are rolled back.
2. The implementation is improved.

File: slice.go [hunk 6]
```diff
@@ -282,6 +282,24 @@ func (d *Decoder) skipSlice(c byte) error {
...
 }
 
+func (d *Decoder) arrayValue(v reflect.Value) error {
+	n, err := d.DecodeBytesLen()
+	if err != nil {
+		return err
+	}
+	if n > v.Len() {
+		n = v.Len()
+	}
```
Comment (vmihailenco) [inline slice.go]: This means that we:
- silently decode only part of the data
- leave some junk in the buffer which can break subsequent `Decode` call

What do you think if we return an error if array does not have enough capacity?

**Comment by h12w** [comment]
1. all bytes are consumed regardless the size of provided array
2. I think partial array should be decoded silently because it is the behavior of "encoding/json"

http://play.golang.org/p/43tPbsLzxZ

In go1.4.3's encoding/json/decode.go, line 437:
```go
		if i < v.Len() {
			// Decode into element.
			d.value(v.Index(i))
		} else {
			// Ran out of fixed array: skip.
			d.value(reflect.Value{})
		}
		i++
```

**Comment by vmihailenco** [comment]
Thanks, merging. I am going to add few tests and benchmark.

**Pull request merged**